# Practical Application III: Comparing Classifiers

**Overview**: In this practical application, your goal is to compare the performance of the classifiers we encountered in this section, namely K Nearest Neighbor, Logistic Regression, Decision Trees, and Support Vector Machines.  We will utilize a dataset related to marketing bank products over the telephone.  



### Getting Started

Our dataset comes from the UCI Machine Learning repository [link](https://archive.ics.uci.edu/ml/datasets/bank+marketing).  The data is from a Portugese banking institution and is a collection of the results of multiple marketing campaigns.  We will make use of the article accompanying the dataset [here](CRISP-DM-BANK.pdf) for more information on the data and features.



### Problem 1: Understanding the Data

To gain a better understanding of the data, please read the information provided in the UCI link above, and examine the **Materials and Methods** section of the paper.  How many marketing campaigns does this data represent?

### Problem 2: Read in the Data

Use pandas to read in the dataset `bank-additional-full.csv` and assign to a meaningful variable name.

In [1]:
import pandas as pd

In [3]:
df = pd.read_csv('data/bank-additional-full.csv', sep = ';')

In [4]:
df.head()

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,56,housemaid,married,basic.4y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
1,57,services,married,high.school,unknown,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
2,37,services,married,high.school,no,yes,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
3,40,admin.,married,basic.6y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
4,56,services,married,high.school,no,no,yes,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no


### Problem 3: Understanding the Features


Examine the data description below, and determine if any of the features are missing values or need to be coerced to a different data type.


```
Input variables:
# bank client data:
1 - age (numeric)
2 - job : type of job (categorical: 'admin.','blue-collar','entrepreneur','housemaid','management','retired','self-employed','services','student','technician','unemployed','unknown')
3 - marital : marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)
4 - education (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')
5 - default: has credit in default? (categorical: 'no','yes','unknown')
6 - housing: has housing loan? (categorical: 'no','yes','unknown')
7 - loan: has personal loan? (categorical: 'no','yes','unknown')
# related with the last contact of the current campaign:
8 - contact: contact communication type (categorical: 'cellular','telephone')
9 - month: last contact month of year (categorical: 'jan', 'feb', 'mar', ..., 'nov', 'dec')
10 - day_of_week: last contact day of the week (categorical: 'mon','tue','wed','thu','fri')
11 - duration: last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y='no'). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.
# other attributes:
12 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)
13 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)
14 - previous: number of contacts performed before this campaign and for this client (numeric)
15 - poutcome: outcome of the previous marketing campaign (categorical: 'failure','nonexistent','success')
# social and economic context attributes
16 - emp.var.rate: employment variation rate - quarterly indicator (numeric)
17 - cons.price.idx: consumer price index - monthly indicator (numeric)
18 - cons.conf.idx: consumer confidence index - monthly indicator (numeric)
19 - euribor3m: euribor 3 month rate - daily indicator (numeric)
20 - nr.employed: number of employees - quarterly indicator (numeric)

Output variable (desired target):
21 - y - has the client subscribed a term deposit? (binary: 'yes','no')
```



In [8]:
# Check for missing values
print("Missing Values:")
print(df.isnull().sum())

Missing Values:
age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
dtype: int64


In [9]:
# Check data types
print("Data Types:")
print(df.dtypes)

Data Types:
age                 int64
job                   str
marital               str
education             str
default               str
housing               str
loan                  str
contact               str
month                 str
day_of_week           str
duration            int64
campaign            int64
pdays               int64
previous            int64
poutcome              str
emp.var.rate      float64
cons.price.idx    float64
cons.conf.idx     float64
euribor3m         float64
nr.employed       float64
y                     str
dtype: object


In [12]:
# Check data shape and detailed info
print("Data Shape:")
print(df.shape)

Data Shape:
(41188, 21)


### Problem 4: Understanding the Task

After examining the description and data, your goal now is to clearly state the *Business Objective* of the task.  State the objective below.

In [13]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  str    
 2   marital         41188 non-null  str    
 3   education       41188 non-null  str    
 4   default         41188 non-null  str    
 5   housing         41188 non-null  str    
 6   loan            41188 non-null  str    
 7   contact         41188 non-null  str    
 8   month           41188 non-null  str    
 9   day_of_week     41188 non-null  str    
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  str    
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null  float64
 1

#### Analysis of Problem 4:

**Business Objective**: Predict whether a customer will subscribe to a term deposit based on their personal information and previous marketing campaign data.

**Key Insight**: This is a binary classification problem where we aim to build predictive models (KNN, Logistic Regression, Decision Trees, SVM) to identify customers most likely to subscribe, enabling targeted marketing efforts.

**Data Context**: The dataset contains 41,188 customer records from a Portuguese banking institution with 20 input features and 1 binary target variable (y: 'yes' or 'no').

In [15]:
# Examine unique values for categorical features to identify any encoding needs
print("Categorical Feature Unique Values:\n")
categorical_cols = df.select_dtypes(include='str').columns
for col in categorical_cols:
    print(f"{col}: {sorted(df[col].unique())}")
    print(f"  Count: {df[col].nunique()} unique values\n")

Categorical Feature Unique Values:

job: ['admin.', 'blue-collar', 'entrepreneur', 'housemaid', 'management', 'retired', 'self-employed', 'services', 'student', 'technician', 'unemployed', 'unknown']
  Count: 12 unique values

marital: ['divorced', 'married', 'single', 'unknown']
  Count: 4 unique values

education: ['basic.4y', 'basic.6y', 'basic.9y', 'high.school', 'illiterate', 'professional.course', 'university.degree', 'unknown']
  Count: 8 unique values

default: ['no', 'unknown', 'yes']
  Count: 3 unique values

housing: ['no', 'unknown', 'yes']
  Count: 3 unique values

loan: ['no', 'unknown', 'yes']
  Count: 3 unique values

contact: ['cellular', 'telephone']
  Count: 2 unique values

month: ['apr', 'aug', 'dec', 'jul', 'jun', 'mar', 'may', 'nov', 'oct', 'sep']
  Count: 10 unique values

day_of_week: ['fri', 'mon', 'thu', 'tue', 'wed']
  Count: 5 unique values

poutcome: ['failure', 'nonexistent', 'success']
  Count: 3 unique values

y: ['no', 'yes']
  Count: 2 unique values



### Problem 5: Engineering Features

Now that you understand your business objective, we will build a basic model to get started.  Before we can do this, we must work to encode the data.  Using just the bank information features, prepare the features and target column for modeling with appropriate encoding and transformations.

#### Feature Selection Rationale

**Selected Features (7 total):**
- **age**: Customer age in years (numeric predictor)
- **job**: Customer occupation (12 categories - strong predictor of subscription likelihood)
- **marital**: Marital status (4 categories - affects decision-making)
- **education**: Educational level (8 categories - correlates with financial behavior)
- **default**: Whether customer has credit in default (3 categories - risk indicator)
- **housing**: Whether customer has a housing loan (3 categories - financial commitment signal)
- **loan**: Whether customer has a personal loan (3 categories - financial commitment signal)

**Excluded Features:**
- **duration**: Last contact duration in seconds - *Critically excluded* because it is only known after the call ends and highly affects the outcome. A realistic model cannot use this for prediction since it's unknown before marketing contact.
- **Economic indicators** (emp.var.rate, cons.price.idx, cons.conf.idx, euribor3m, nr.employed): These are macro-economic factors that change monthly but are less actionable for targeted marketing campaigns. We focus on individual customer characteristics instead.
- **Contact metadata** (contact, month, day_of_week, pdays, previous, campaign, poutcome): These campaign-specific features will be analyzed separately if needed, but we start with stable customer characteristics.

**Why These Features?**
This core set focuses on *stable, actionable customer attributes* that bank employees would know during initial contact. These features are sufficient to build an initial baseline model and understand which customer characteristics are most predictive of subscription behavior.

In [16]:
# Step 1: Identify bank information features (exclude duration as it's not known before the call)
# Bank client data features
bank_features = ['age', 'job', 'marital', 'education', 'default', 'housing', 'loan']

# Create feature matrix X and target vector y
X = df[bank_features].copy()
y = df['y'].copy()

print("Features selected for modeling:")
print(X.columns.tolist())
print(f"\nFeature matrix shape: {X.shape}")
print(f"Target vector shape: {y.shape}")
print(f"\nFeature data types:\n{X.dtypes}")

Features selected for modeling:
['age', 'job', 'marital', 'education', 'default', 'housing', 'loan']

Feature matrix shape: (41188, 7)
Target vector shape: (41188,)

Feature data types:
age          int64
job            str
marital        str
education      str
default        str
housing        str
loan           str
dtype: object


In [17]:
# Step 2: Encode categorical variables using one-hot encoding
from sklearn.preprocessing import LabelEncoder

# One-hot encode categorical features
X_encoded = pd.get_dummies(X, columns=['job', 'marital', 'education', 'default', 'housing', 'loan'], 
                            drop_first=False)

# Encode target variable: 'yes' -> 1, 'no' -> 0
y_encoded = LabelEncoder().fit_transform(y)

print("Encoded features shape:", X_encoded.shape)
print(f"Target encoding: 'no' -> 0, 'yes' -> 1")
print(f"Target value distribution:\n{pd.Series(y_encoded).value_counts().sort_index()}")
print(f"\nEncoded feature columns ({len(X_encoded.columns)} total):")
print(X_encoded.columns.tolist())

Encoded features shape: (41188, 34)
Target encoding: 'no' -> 0, 'yes' -> 1
Target value distribution:
0    36548
1     4640
Name: count, dtype: int64

Encoded feature columns (34 total):
['age', 'job_admin.', 'job_blue-collar', 'job_entrepreneur', 'job_housemaid', 'job_management', 'job_retired', 'job_self-employed', 'job_services', 'job_student', 'job_technician', 'job_unemployed', 'job_unknown', 'marital_divorced', 'marital_married', 'marital_single', 'marital_unknown', 'education_basic.4y', 'education_basic.6y', 'education_basic.9y', 'education_high.school', 'education_illiterate', 'education_professional.course', 'education_university.degree', 'education_unknown', 'default_no', 'default_unknown', 'default_yes', 'housing_no', 'housing_unknown', 'housing_yes', 'loan_no', 'loan_unknown', 'loan_yes']


In [18]:
# Step 3: Display summary of prepared data
print("Data Preparation Summary:")
print("=" * 60)
print(f"\nOriginal dataset: {df.shape[0]} rows × {df.shape[1]} columns")
print(f"Selected features: {len(bank_features)} features")
print(f"Encoded features: {X_encoded.shape[1]} features (after one-hot encoding)")
print(f"\nFeatures breakdown:")
print(f"  - Numeric: age (1)")
print(f"  - Categorical encoded: job, marital, education, default, housing, loan")
print(f"\nTarget variable:")
print(f"  - Class 0 (no): {(y_encoded == 0).sum()} samples ({(y_encoded == 0).sum()/len(y_encoded)*100:.1f}%)")
print(f"  - Class 1 (yes): {(y_encoded == 1).sum()} samples ({(y_encoded == 1).sum()/len(y_encoded)*100:.1f}%)")
print(f"\nData is ready for modeling!")
print(f"X_encoded shape: {X_encoded.shape}")
print(f"y_encoded shape: {y_encoded.shape}")

Data Preparation Summary:

Original dataset: 41188 rows × 21 columns
Selected features: 7 features
Encoded features: 34 features (after one-hot encoding)

Features breakdown:
  - Numeric: age (1)
  - Categorical encoded: job, marital, education, default, housing, loan

Target variable:
  - Class 0 (no): 36548 samples (88.7%)
  - Class 1 (yes): 4640 samples (11.3%)

Data is ready for modeling!
X_encoded shape: (41188, 34)
y_encoded shape: (41188,)


### Problem 6: Train/Test Split

With your data prepared, split it into a train and test set.

### Problem 7: A Baseline Model

Before we build our first model, we want to establish a baseline.  What is the baseline performance that our classifier should aim to beat?

### Problem 8: A Simple Model

Use Logistic Regression to build a basic model on your data.  

### Problem 9: Score the Model

What is the accuracy of your model?

### Problem 10: Model Comparisons

Now, we aim to compare the performance of the Logistic Regression model to our KNN algorithm, Decision Tree, and SVM models.  Using the default settings for each of the models, fit and score each.  Also, be sure to compare the fit time of each of the models.  Present your findings in a `DataFrame` similar to that below:

| Model | Train Time | Train Accuracy | Test Accuracy |
| ----- | ---------- | -------------  | -----------   |
|     |    |.     |.     |

### Problem 11: Improving the Model

Now that we have some basic models on the board, we want to try to improve these.  Below, we list a few things to explore in this pursuit.


- Hyperparameter tuning and grid search.  All of our models have additional hyperparameters to tune and explore.  For example the number of neighbors in KNN or the maximum depth of a Decision Tree.  
- Adjust your performance metric

##### Questions